File: python/sentroy/inbox.py
from __future__ import annotations

from typing import Any, Optional

from sentroy._http import _HttpClient
from sentroy.types import (
    InboxListParams,
    Mailbox,
    MessageDetail,
    MessageSummary,
)


class InboxResource:
    """Interact with the Inbox API."""

    def __init__(self, http: _HttpClient) -> None:
        self._http = http

    def list(self, params: Optional[InboxListParams] = None) -> list[MessageSummary]:
        """List messages in a mailbox folder."""
        query: dict[str, Any] = {}
        if params is not None:
            if params.mailbox is not None:
                query["mailbox"] = params.mailbox
            if params.folder is not None:
                query["folder"] = params.folder
            if params.page is not None:
                query["page"] = params.page
            if params.limit is not None:
                query["limit"] = params.limit
            if params.unread_only:
                query["unreadOnly"] = "true"
        data = self._http.get("/inbox", query or None)
        return [MessageSummary.from_dict(d) for d in (data or [])]

    def get(
        self,
        uid: int,
        *,
        mailbox: Optional[str] = None,
        folder: Optional[str] = None,
    ) -> MessageDetail:
        """Get a single message detail."""
        query: dict[str, Any] = {}
        if mailbox is not None:
            query["mailbox"] = mailbox
        if folder is not None:
            query["folder"] = folder
        data = self._http.get(f"/inbox/{uid}", query or None)
        return MessageDetail.from_dict(data)

    def list_folders(self, mailbox: Optional[str] = None) -> list[Mailbox]:
        """List IMAP folders (mailboxes) for a given email account."""
        query: dict[str, Any] = {}
        if mailbox is not None:
            query["mailbox"] = mailbox
        data = self._http.get("/inbox/mailboxes", query or None)
        return [Mailbox.from_dict(d) for d in (data or [])]

    def get_thread(
        self,
        subject: str,
        mailbox: Optional[str] = None,
    ) -> list[MessageDetail]:
        """Get thread messages by subject."""
        query: dict[str, Any] = {"subject": subject}
        if mailbox is not None:
            query["mailbox"] = mailbox
        data = self._http.get("/inbox/thread", query)
        return [MessageDetail.from_dict(d) for d in (data or [])]

    def mark_as_read(
        self,
        uid: int,
        *,
        mailbox: Optional[str] = None,
        folder: Optional[str] = None,
    ) -> None:
        """Mark a message as read."""
        self._http.post(f"/inbox/{uid}/read", {
            "mailbox": mailbox,
            "folder": folder,
        })

    def mark_as_unread(
        self,
        uid: int,
        *,
        mailbox: Optional[str] = None,
        folder: Optional[str] = None,
    ) -> None:
        """Mark a message as unread."""
        self._http.delete(f"/inbox/{uid}/read", {
            "mailbox": mailbox,
            "folder": folder,
        })

    def move(
        self,
        uid: int,
        to: str,
        *,
        from_folder: Optional[str] = None,
        mailbox: Optional[str] = None,
    ) -> None:
        """Move a message to another folder."""
        self._http.post(f"/inbox/{uid}/move", {
            "to": to,
            "from": from_folder,
            "mailbox": mailbox,
        })

    def delete(
        self,
        uid: int,
        *,
        mailbox: Optional[str] = None,
        folder: Optional[str] = None,
    ) -> None:
        """Delete a message."""
        query: dict[str, Any] = {}
        if mailbox is not None:
            query["mailbox"] = mailbox
        if folder is not None:
            query["folder"] = folder
        self._http.delete(f"/inbox/{uid}", query or None)

File: typescript/src/resources/inbox.ts
import type { HttpClient } from "../http"
import type {
  MessageSummary,
  MessageDetail,
  Mailbox,
  InboxListParams,
} from "../types"

export class Inbox {
  constructor(private http: HttpClient) {}

  /** List messages in a mailbox folder */
  async list(params?: InboxListParams): Promise<MessageSummary[]> {
    const query: Record<string, unknown> = {}
    if (params?.mailbox) query.mailbox = params.mailbox
    if (params?.folder) query.folder = params.folder
    if (params?.page) query.page = params.page
    if (params?.limit) query.limit = params.limit
    if (params?.unreadOnly) query.unreadOnly = "true"
    return this.http.get<MessageSummary[]>("/inbox", query)
  }

  /** Get a single message detail */
  async get(
    uid: number,
    options?: { mailbox?: string; folder?: string },
  ): Promise<MessageDetail> {
    const query: Record<string, unknown> = {}
    if (options?.mailbox) query.mailbox = options.mailbox
    if (options?.folder) query.folder = options.folder
    return this.http.get<MessageDetail>(`/inbox/${uid}`, query)
  }

  /** List IMAP folders (mailboxes) for a given email account */
  async listFolders(mailbox?: string): Promise<Mailbox[]> {
    const query: Record<string, unknown> = {}
    if (mailbox) query.mailbox = mailbox
    return this.http.get<Mailbox[]>("/inbox/mailboxes", query)
  }

  /** Get thread messages by subject */
  async getThread(
    subject: string,
    mailbox?: string,
  ): Promise<(MessageDetail & { folder: string })[]> {
    const query: Record<string, unknown> = { subject }
    if (mailbox) query.mailbox = mailbox
    return this.http.get<(MessageDetail & { folder: string })[]>(
      "/inbox/thread",
      query,
    )
  }

  /** Mark a message as read */
  async markAsRead(
    uid: number,
    options?: { mailbox?: string; folder?: string },
  ): Promise<void> {
    await this.http.post(`/inbox/${uid}/read`, {
      mailbox: options?.mailbox,
      folder: options?.folder,
    })
  }

  /** Mark a message as unread */
  async markAsUnread(
    uid: number,
    options?: { mailbox?: string; folder?: string },
  ): Promise<void> {
    await this.http.del(`/inbox/${uid}/read`, {
      mailbox: options?.mailbox,
      folder: options?.folder,
    })
  }

  /** Move a message to another folder */
  async move(
    uid: number,
    to: string,
    options?: { from?: string; mailbox?: string },
  ): Promise<void> {
    await this.http.post(`/inbox/${uid}/move`, {
      to,
      from: options?.from,
      mailbox: options?.mailbox,
    })
  }

  /** Delete a message */
  async delete(
    uid: number,
    options?: { mailbox?: string; folder?: string },
  ): Promise<void> {
    const query: Record<string, unknown> = {}
    if (options?.mailbox) query.mailbox = options.mailbox
    if (options?.folder) query.folder = options.folder
    await this.http.del(`/inbox/${uid}`, query)
  }
}
